packages/server-boilerplate/src/utils/handleWith.ts

```diff
@@ -0,0 +1,16 @@
+/*
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ *
+ */
+import { Response, NextFunction } from 'express';
+import { Route } from '../routes';
+
+type ReqOfRoute<T> = T extends Route<infer Req> ? Req : never;
+// type ResOfRoute<T> = T extends Route<any, infer Res> ? Res : never;
+export const handleWith = <T extends { handle: () => Promise<void> }>(
+  RouteClass: new (req: ReqOfRoute<T>, res: Response, next: NextFunction) => T,
+) => (req: ReqOfRoute<T>, res: Response, next: NextFunction) => {
```

Comment:
More generics madness 😝 I think I have a problem...

`T` here is a Route subtype, eg. `LoginRoute`. The `new (...) => T` syntax is pretty magical, it implies that the parameter is a class which when instantiated creates a Route of type `T`.

Reply:
Nice one! This syntax seems pretty handy. Thanks for the explanation!